packages/workbox-routing/lib/Router.mjs

```diff
@@ -0,0 +1,248 @@
+ */
+class Router {
+  /**
+   * Constructs a new `Router` instance, without any registered routes.
+   */
+  constructor() {
+    // _routes will contain a mapping of HTTP method name ('GET', etc.) to an
+    // array of all the corresponding Route instances that are registered.
+    this._routes = new Map();
+    this._isListenerRegistered = false;
+  }
+
+  /**
+   * Apply the routing rules to a FetchEvent object to get a Response from an
+   * appropriate handler.
+   *
+   * @param {FetchEvent} event The event passed in to a `fetch` handler.
+   * @return {Promise<Response>|undefined} Returns a promise for a response,
+   * taking the registered routes into account. If there was no matching route
+   * and there's no `defaultHandler`, then returns undefined.
+   */
+  handleRequest(event) {
+    if (process.env.NODE_ENV !== 'production') {
+      core.assert.isInstance(event, FetchEvent, {
+        moduleName: 'workbox-routing',
+        className: 'Router',
+        funcName: 'handleRequest',
+        paramName: 'event',
+      });
+    }
+
+    const url = new URL(event.request.url);
+    if (!url.protocol.startsWith('http')) {
+      if (process.env.NODE_ENV !== 'production') {
+        _private.logger.log(`The URL '${url}' does not start with 'http', so ` +
+          `it can't be handled.`);
+      }
+      return;
+    }
+
+    let {handler, params} = this._findHandlerAndParams(event, url);
+
+    // If we don't have a handler because there was no matching route, then
+    // fall back to defaultHandler if that's defined.
+    if (!handler && this.defaultHandler) {
+      handler = this.defaultHandler;
+    }
+
+    if (handler) {
+      let responsePromise = handler.handle({url, event, params});
+      if (this.catchHandler) {
+        responsePromise = responsePromise.catch((error) => {
+          return this.catchHandler.handle({url, event, error});
+        });
+      }
+      return responsePromise;
+    }
+  }
+
+  /**
+   * Checks the incoming event.request against the registered routes, and if
+   * there's a match, returns the corresponding handler along with any params
+   * generated by the match.
+   *
+   * @param {FetchEvent} event
+   * @param {URL} url
+   * @return {Object} Returns an object with `handler` and `params` properties.
+   * They are populated if a matching route was found or `undefined` otherwise.
+   * @private
+   */
+  _findHandlerAndParams(event, url) {
+    const routes = this._routes.get(event.request.method) || [];
+    for (const route of routes) {
+      let matchResult = route.match({url, event});
+      if (matchResult) {
+        _private.logger.log(`The router found a matching route.`, {
+          route,
+          request: event.request,
+        });
+
+        if (Array.isArray(matchResult) && matchResult.length === 0) {
+          // Instead of passing an empty array in as params, use undefined.
+          matchResult = undefined;
+        } else if (matchResult.constructor === Object &&
```

Comment:
Is there a reason to check `matchResult.constructor` vs typeof matchResult === 'Object' ? The latter is a bit clearer in terms of intention.

Reply:
We let the match function return whatever it wants, and `typeof SomeClassInstance` will also return `'object'`. This special-case stuff is needed due to ExpressRoute commonly returning these messy values.